File: python/src/python/models/WithdrawEvent.py
from dataclasses import dataclass, field, asdict
from typing import Optional, Dict, Any, List
import re
from datetime import datetime


@dataclass
class WithdrawEvent:
    account_id: Optional[str] = None
    workspace_id: Optional[str] = None
    user_id: Optional[str] = None
    event_category: str = "Withdraw"
    event_name: Optional[str] = None
    event_id: Optional[str] = None
    event_time: Optional[str] = None
    amount: Optional[float] = None
    payment_method: Optional[str] = None
    transaction_id: Optional[str] = None
    failure_reason: Optional[str] = None

    # -------------------------------
    # Validation methods
    # -------------------------------
    def validate(self) -> Dict[str, Any]:
        errors: List[str] = []

        # Required fields
        if not self.account_id:
            errors.append("account_id is required")
        if not self.workspace_id:
            errors.append("workspace_id is required")
        if not self.user_id:
            errors.append("user_id is required")
        if not self.event_name:
            errors.append("event_name is required")
        if not self.event_id:
            errors.append("event_id is required")
        if not self.event_time:
            errors.append("event_time is required")
        if self.amount is None:
            errors.append("amount is required")
        if not self.payment_method:
            errors.append("payment_method is required")
        if not self.transaction_id:
            errors.append("transaction_id is required")

        # Event category validation
        if self.event_category and self.event_category != "Withdraw":
            errors.append('event_category must be "Withdraw" for withdraw events')

        # Event name validation
        valid_event_names = [
            "Successful Withdrawal",
            "Failed Withdrawal",
            "Pending Withdrawal",
            "Withdrawal Cancelled",
            "Withdrawal Rejected"
        ]
        if self.event_name and self.event_name not in valid_event_names:
            errors.append(f"event_name must be one of: {', '.join(valid_event_names)}")

        # Amount validation
        if self.amount is not None and self.amount <= 0:
            errors.append("amount must be a positive number")

        # Payment method validation
        valid_payment_methods = [
            "bank", "credit_card", "debit_card", "e_wallet",
            "crypto", "paypal", "skrill", "neteller"
        ]
        if self.payment_method and self.payment_method not in valid_payment_methods:
            errors.append(f"payment_method must be one of: {', '.join(valid_payment_methods)}")


        # Date format validation
        if self.event_time and not self.is_valid_datetime(self.event_time):
            errors.append("event_time must be in ISO 8601 format (YYYY-MM-DDTHH:mm:ssZ)")

        return {"isValid": len(errors) == 0, "errors": errors}

    @staticmethod
    def is_valid_datetime(date_time: str) -> bool:
        try:
            datetime.fromisoformat(date_time.replace("Z", "+00:00"))
            return True
        except ValueError:
            return False

    # -------------------------------
    # Utility methods
    # -------------------------------
    def to_dict(self) -> Dict[str, Any]:
        return {k: v for k, v in asdict(self).items() if v is not None}

    @classmethod
    def from_object(cls, data: Dict[str, Any]) -> "WithdrawEvent":
        return cls(**data)

    def __repr__(self) -> str:
        return f"WithdrawEvent(event_id={self.event_id}, event_name={self.event_name}, user_id={self.user_id})"

File: js/src/models/WithdrawEvent.js
/**
 * Withdraw Event Model
 * Represents withdrawal-related events for the Data Pipeline API
 */
class WithdrawEvent {
  constructor(data = {}) {
    this.account_id = data.account_id;
    this.workspace_id = data.workspace_id;
    this.user_id = data.user_id;
    this.event_category = data.event_category || 'Withdraw';
    this.event_name = data.event_name;
    this.event_id = data.event_id;
    this.event_time = data.event_time;
    this.amount = data.amount;
    this.payment_method = data.payment_method;
    this.transaction_id = data.transaction_id;
    this.status = data.status;
    this.currency = data.currency;
    this.fees = data.fees;
    this.net_amount = data.net_amount;
    this.withdrawal_reason = data.withdrawal_reason;
    this.processing_time = data.processing_time;
    this.failure_reason = data.failure_reason;
  }

  /**
   * Validates the withdraw event data
   * @returns {Object} Validation result with isValid boolean and errors array
   */
  validate() {
    const errors = [];

    // Required fields
    if (!this.account_id) errors.push('account_id is required');
    if (!this.workspace_id) errors.push('workspace_id is required');
    if (!this.user_id) errors.push('user_id is required');
    if (!this.event_name) errors.push('event_name is required');
    if (!this.event_id) errors.push('event_id is required');
    if (!this.event_time) errors.push('event_time is required');
    if (!this.amount) errors.push('amount is required');
    if (!this.payment_method) errors.push('payment_method is required');
    if (!this.transaction_id) errors.push('transaction_id is required');

    // Event category validation
    if (this.event_category && this.event_category !== 'Withdraw') {
      errors.push('event_category must be "Withdraw" for withdraw events');
    }

    // Event name validation
    const validEventNames = [
      'Successful Withdrawal',
      'Failed Withdrawal',
      'Pending Withdrawal',
      'Withdrawal Cancelled',
      'Withdrawal Rejected'
    ];

    if (this.event_name && !validEventNames.includes(this.event_name)) {
      errors.push(`event_name must be one of: ${validEventNames.join(', ')}`);
    }

    // Amount validation
    if (this.amount && (typeof this.amount !== 'number' || this.amount <= 0)) {
      errors.push('amount must be a positive number');
    }

    // Payment method validation
    const validPaymentMethods = [
      'bank',
      'credit_card',
      'debit_card',
      'e_wallet',
      'crypto',
      'paypal',
      'skrill',
      'neteller'
    ];

    if (this.payment_method && !validPaymentMethods.includes(this.payment_method)) {
      errors.push(`payment_method must be one of: ${validPaymentMethods.join(', ')}`);
    }

    // Status validation
    if (this.status && !['success', 'pending', 'failed', 'cancelled', 'rejected'].includes(this.status)) {
      errors.push('status must be one of: success, pending, failed, cancelled, rejected');
    }

    // Currency validation
    if (this.currency && !this.isValidCurrency(this.currency)) {
      errors.push('currency must be a valid 3-letter ISO currency code');
    }

    // Date format validation
    if (this.event_time && !this.isValidDateTime(this.event_time)) {
      errors.push('event_time must be in ISO 8601 format (YYYY-MM-DDTHH:mm:ssZ)');
    }

    return {
      isValid: errors.length === 0,
      errors
    };
  }

  /**
   * Validates ISO 8601 datetime format
   * @param {string} dateTime - DateTime string to validate
   * @returns {boolean} True if datetime is valid
   */
  isValidDateTime(dateTime) {
    const date = new Date(dateTime);
    return date instanceof Date && !isNaN(date);
  }

  /**
   * Validates currency code format
   * @param {string} currency - Currency code to validate
   * @returns {boolean} True if currency code is valid
   */
  isValidCurrency(currency) {
    const currencyRegex = /^[A-Z]{3}$/;
    return currencyRegex.test(currency);
  }

  /**
   * Converts the model to a plain object
   * @returns {Object} Plain object representation
   */
  toJSON() {
    const obj = {};
    Object.keys(this).forEach(key => {
      if (this[key] !== undefined) {
        obj[key] = this[key];
      }
    });
    return obj;
  }

  /**
   * Creates a WithdrawEvent instance from plain object
   * @param {Object} data - Plain object data
   * @returns {WithdrawEvent} New WithdrawEvent instance
   */
  static fromObject(data) {
    return new WithdrawEvent(data);
  }
}

module.exports = WithdrawEvent;

